deployer/validation.go

```diff
@@ -0,0 +1,211 @@
+	c, err := net.Dial("tcp", fmt.Sprintf("%s:%d", listenIP, port))
+	if err != nil {
+		return errors.Wrapf(err, "failed to connect to ip. make sure the twin id is associated with a valid yggdrasil ip, twin id: %d, ygg ip: %s, err", tfPluginClient.TwinID, yggIP)
+	}
+	c.Close()
+	if !arrived {
+		return errors.Wrapf(err, "sent request but didn't arrive to me. make sure the twin id is associated with a valid yggdrasil ip, twin id: %d, ygg ip: %s, err", tfPluginClient.TwinID, yggIP)
+	}
+	return nil
+}
+
+func validateRMB(tfPluginClient *TFPluginClient, sub subi.SubstrateExt) error {
+	if err := validateRedis(tfPluginClient); err != nil {
+		return err
+	}
+
+	return validateYggdrasil(tfPluginClient, sub)
+}
+
+func validateRMBProxyServer(tfPluginClient *TFPluginClient) error {
+	return tfPluginClient.GridProxyClient.Ping()
+}
+
+func validateMnemonics(mnemonics string) error {
+	if len(mnemonics) == 0 {
+		return errors.New("mnemonics required")
+	}
+
+	alphaOnly := regexp.MustCompile(`^[a-zA-Z\s]+$`)
+	if !alphaOnly.MatchString(mnemonics) {
+		return errors.New("mnemonics can only be composed of a non-alphanumeric character or a whitespace")
+	}
+
+	return nil
+}
+
+func validateSubstrateURL(url string) error {
+	if len(url) == 0 {
+		return errors.New("substrate url is required")
+	}
+
+	alphaOnly := regexp.MustCompile(`^wss:\/\/[a-z0-9]+\.[a-z0-9]\/?([^\s<>\#%"\,\{\}\\|\\\^\[\]]+)?$`)
+	if !alphaOnly.MatchString(url) {
+		return errors.New("substrate url is not valid")
+	}
+
+	return nil
+}
+
+func validateProxyURL(url string) error {
+	if len(url) == 0 {
+		return errors.New("proxy url is required")
+	}
+
+	alphaOnly := regexp.MustCompile(`^https:\/\/[a-z0-9]+\.[a-z0-9]\/?([^\s<>\#%"\,\{\}\\|\\\^\[\]]+)?$`)
+	if !alphaOnly.MatchString(url) {
+		return errors.New("proxy url is not valid")
+	}
+
+	return nil
+}
+
+func validateClientRMB(tfPluginClient *TFPluginClient, sub subi.SubstrateExt) error {
+	if tfPluginClient.UseRmbProxy {
+		return validateRMBProxyServer(tfPluginClient)
+	}
+
+	return validateRMB(tfPluginClient, sub)
+}
+
+func validateAccountBalanceForExtrinsics(sub subi.SubstrateExt, identity substrate.Identity) error {
+	balance, err := sub.GetBalance(identity)
+	if err != nil && !errors.Is(err, substrate.ErrAccountNotFound) {
+		return errors.Wrap(err, "failed to get account with the given mnemonics")
+	}
+	log.Printf("money %d\n", balance.Free)
+	if balance.Free.Cmp(big.NewInt(20000)) == -1 {
+		return fmt.Errorf("account contains %s, min fee is 20000", balance.Free)
+	}
+	return nil
+}
+
+func newRedisPool(address string) (*redis.Pool, error) {
+	u, err := url.Parse(address)
+	if err != nil {
+		return nil, err
+	}
+	var host string
+	switch u.Scheme {
+	case "tcp":
+		host = u.Host
+	case "unix":
+		host = u.Path
+	default:
+		return nil, fmt.Errorf("unknown scheme '%s' expecting tcp or unix", u.Scheme)
+	}
+	var opts []redis.DialOption
+
+	if u.User != nil {
+		opts = append(
+			opts,
+			redis.DialPassword(u.User.Username()),
```

Comment:
apparently it is done that way because the url used is like `redis://<password>@address:port` without the user part so `u.User.Username()` actually returns the `<password>` part because there is no user